File: packages/go-server/override_test.go
package savvagent

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestConfigurationOverrides(t *testing.T) {
	t.Run("Override configuration completely when merge is false", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"primaryColor": "#007bff",
				"fontSize":     float64(16),
			},
			"limits": map[string]interface{}{
				"maxItems": float64(100),
			},
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		// Set override (merge = false by default)
		overrideConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"primaryColor": "#ff0000",
			},
		}
		client.SetConfigOverride("test-flag", overrideConfig, nil)

		result, err := client.Evaluate("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if !result.Value {
			t.Error("Expected flag to be enabled")
		}

		theme := result.Configuration["theme"].(map[string]interface{})
		if theme["primaryColor"] != "#ff0000" {
			t.Errorf("Expected primaryColor to be overridden")
		}
		if _, exists := theme["fontSize"]; exists {
			t.Error("fontSize should not exist in full override")
		}
	})

	t.Run("Merge configuration when merge is true", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"primaryColor": "#007bff",
				"fontSize":     float64(16),
			},
			"limits": map[string]interface{}{
				"maxItems": float64(100),
			},
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		// Set override with merge = true
		overrideConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"primaryColor": "#ff0000",
			},
			"newField": "added",
		}
		client.SetConfigOverride("test-flag", overrideConfig, &ConfigOverrideOptions{Merge: true})

		result, err := client.Evaluate("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}

		theme := result.Configuration["theme"].(map[string]interface{})
		if theme["primaryColor"] != "#ff0000" {
			t.Error("Expected primaryColor to be overridden")
		}
		if theme["fontSize"] != float64(16) {
			t.Error("Expected fontSize to be preserved from API")
		}
		if result.Configuration["newField"] != "added" {
			t.Error("Expected newField to be added")
		}
		if result.Configuration["limits"] == nil {
			t.Error("Expected limits to be preserved from API")
		}
	})

	t.Run("Override variation", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":     true,
				"variation": "control",
				"configuration": map[string]interface{}{
					"algorithm": "standard",
				},
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		// Set variation override
		client.SetVariationOverride("test-flag", "variant_b")

		result, err := client.Evaluate("test-flag", nil)

		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if result.Variation != "variant_b" {
			t.Errorf("Expected variation 'variant_b', got '%s'", result.Variation)
		}
	})

	t.Run("Clear configuration override", func(t *testing.T) {
		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value": true,
				"configuration": map[string]interface{}{
					"original": "config",
				},
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		// Set override
		client.SetConfigOverride("test-flag", map[string]interface{}{
			"overridden": "value",
		}, nil)

		// First call should have override
		result1, _ := client.Evaluate("test-flag", nil)
		if result1.Configuration["overridden"] != "value" {
			t.Error("Expected overridden configuration")
		}

		// Clear override
		client.ClearConfigOverride("test-flag")

		// Second call should have original config
		result2, _ := client.Evaluate("test-flag", nil)
		if result2.Configuration["original"] != "config" {
			t.Error("Expected original configuration")
		}
	})

	t.Run("Check if override exists", func(t *testing.T) {
		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: "https://api.test.com",
		})

		if client.HasConfigOverride("test-flag") {
			t.Error("Expected no override initially")
		}

		client.SetConfigOverride("test-flag", map[string]interface{}{"test": "value"}, nil)

		if !client.HasConfigOverride("test-flag") {
			t.Error("Expected override to exist")
		}

		client.ClearConfigOverride("test-flag")

		if client.HasConfigOverride("test-flag") {
			t.Error("Expected override to be cleared")
		}
	})

	t.Run("Clear all overrides", func(t *testing.T) {
		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: "https://api.test.com",
		})

		client.SetConfigOverride("flag1", map[string]interface{}{"config": "a"}, nil)
		client.SetVariationOverride("flag2", "variant_a")

		if !client.HasConfigOverride("flag1") || !client.HasVariationOverride("flag2") {
			t.Error("Expected overrides to exist")
		}

		client.ClearAllOverrides()

		if client.HasConfigOverride("flag1") || client.HasVariationOverride("flag2") {
			t.Error("Expected all overrides to be cleared")
		}
	})

	t.Run("Deep merge nested configurations", func(t *testing.T) {
		mockConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"colors": map[string]interface{}{
					"primary":   "#007bff",
					"secondary": "#6c757d",
				},
				"fontSize": float64(16),
			},
		}

		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			response := map[string]interface{}{
				"value":         true,
				"configuration": mockConfig,
			}
			json.NewEncoder(w).Encode(response)
		}))
		defer server.Close()

		client, _ := NewClient(Config{
			APIKey:  "sdk_test_key",
			BaseURL: server.URL,
		})

		// Override nested configuration
		overrideConfig := map[string]interface{}{
			"theme": map[string]interface{}{
				"colors": map[string]interface{}{
					"primary": "#ff0000",
				},
			},
		}
		client.SetConfigOverride("test-flag", overrideConfig, &ConfigOverrideOptions{Merge: true})

		result, _ := client.Evaluate("test-flag", nil)

		theme := result.Configuration["theme"].(map[string]interface{})
		colors := theme["colors"].(map[string]interface{})

		if colors["primary"] != "#ff0000" {
			t.Error("Expected primary color to be overridden")
		}
		if colors["secondary"] != "#6c757d" {
			t.Error("Expected secondary color to be preserved")
		}
		if theme["fontSize"] != float64(16) {
			t.Error("Expected fontSize to be preserved")
		}
	})
}

File: packages/rust-server/tests/override_test.rs
use savvagent_sdk::{Config, ConfigOverrideOptions, FlagClient};
use serde_json::json;
use wiremock::{Mock, MockServer, ResponseTemplate};
use wiremock::matchers::{method, path};

#[tokio::test]
async fn test_override_configuration_completely() {
    let mock_server = MockServer::start().await;

    let config_json = json!({
        "theme": {
            "primaryColor": "#007bff",
            "fontSize": 16
        },
        "limits": {
            "maxItems": 100
        }
    });

    Mock::given(method("POST"))
        .and(path("/api/evaluate/test-flag"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "value": true,
            "configuration": config_json
        })))
        .mount(&mock_server)
        .await;

    let config = Config::new("sdk_test_key")
        .with_base_url(mock_server.uri());

    let client = FlagClient::new(config).unwrap();

    // Set override (merge = false by default)
    client.set_config_override(
        "test-flag",
        json!({
            "theme": {
                "primaryColor": "#ff0000"
            }
        }),
        None,
    ).unwrap();

    let result = client.evaluate("test-flag", None).await.unwrap();

    assert!(result.value);
    assert!(result.configuration.is_some());

    let config = result.configuration.unwrap();
    assert_eq!(config["theme"]["primaryColor"], "#ff0000");
    assert!(config["theme"].get("fontSize").is_none());
}

#[tokio::test]
async fn test_merge_configuration() {
    let mock_server = MockServer::start().await;

    let config_json = json!({
        "theme": {
            "primaryColor": "#007bff",
            "fontSize": 16
        },
        "limits": {
            "maxItems": 100
        }
    });

    Mock::given(method("POST"))
        .and(path("/api/evaluate/test-flag"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "value": true,
            "configuration": config_json
        })))
        .mount(&mock_server)
        .await;

    let config = Config::new("sdk_test_key")
        .with_base_url(mock_server.uri());

    let client = FlagClient::new(config).unwrap();

    // Set override with merge = true
    client.set_config_override(
        "test-flag",
        json!({
            "theme": {
                "primaryColor": "#ff0000"
            },
            "newField": "added"
        }),
        Some(ConfigOverrideOptions::new().with_merge(true)),
    ).unwrap();

    let result = client.evaluate("test-flag", None).await.unwrap();

    assert!(result.value);

    let config = result.configuration.unwrap();
    assert_eq!(config["theme"]["primaryColor"], "#ff0000");
    assert_eq!(config["theme"]["fontSize"], 16);
    assert_eq!(config["newField"], "added");
    assert!(config.get("limits").is_some());
}

#[tokio::test]
async fn test_override_variation() {
    let mock_server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/api/evaluate/test-flag"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "value": true,
            "variation": "control",
            "configuration": {
                "algorithm": "standard"
            }
        })))
        .mount(&mock_server)
        .await;

    let config = Config::new("sdk_test_key")
        .with_base_url(mock_server.uri());

    let client = FlagClient::new(config).unwrap();

    // Set variation override
    client.set_variation_override("test-flag", "variant_b");

    let result = client.evaluate("test-flag", None).await.unwrap();

    assert!(result.value);
    assert_eq!(result.variation.unwrap(), "variant_b");
}

#[tokio::test]
async fn test_clear_configuration_override() {
    let mock_server = MockServer::start().await;

    Mock::given(method("POST"))
        .and(path("/api/evaluate/test-flag"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "value": true,
            "configuration": {
                "original": "config"
            }
        })))
        .mount(&mock_server)
        .await;

    let config = Config::new("sdk_test_key")
        .with_base_url(mock_server.uri());

    let client = FlagClient::new(config).unwrap();

    // Set override
    client.set_config_override(
        "test-flag",
        json!({ "overridden": "value" }),
        None,
    ).unwrap();

    // First call should have override
    let result1 = client.evaluate("test-flag", None).await.unwrap();
    assert_eq!(result1.configuration.unwrap()["overridden"], "value");

    // Clear override
    client.clear_config_override("test-flag");

    // Second call should have original config
    let result2 = client.evaluate("test-flag", None).await.unwrap();
    assert_eq!(result2.configuration.unwrap()["original"], "config");
}

#[tokio::test]
async fn test_has_override() {
    let config = Config::new("sdk_test_key");
    let client = FlagClient::new(config).unwrap();

    assert!(!client.has_config_override("test-flag"));

    client.set_config_override(
        "test-flag",
        json!({ "test": "value" }),
        None,
    ).unwrap();

    assert!(client.has_config_override("test-flag"));

    client.clear_config_override("test-flag");

    assert!(!client.has_config_override("test-flag"));
}

#[tokio::test]
async fn test_clear_all_overrides() {
    let config = Config::new("sdk_test_key");
    let client = FlagClient::new(config).unwrap();

    client.set_config_override("flag1", json!({ "config": "a" }), None).unwrap();
    client.set_variation_override("flag2", "variant_a");

    assert!(client.has_config_override("flag1"));
    assert!(client.has_variation_override("flag2"));

    client.clear_all_overrides();

    assert!(!client.has_config_override("flag1"));
    assert!(!client.has_variation_override("flag2"));
}

#[tokio::test]
async fn test_deep_merge_nested_configurations() {
    let mock_server = MockServer::start().await;

    let config_json = json!({
        "theme": {
            "colors": {
                "primary": "#007bff",
                "secondary": "#6c757d"
            },
            "fontSize": 16
        }
    });

    Mock::given(method("POST"))
        .and(path("/api/evaluate/test-flag"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "value": true,
            "configuration": config_json
        })))
        .mount(&mock_server)
        .await;

    let config = Config::new("sdk_test_key")
        .with_base_url(mock_server.uri());

    let client = FlagClient::new(config).unwrap();

    // Override nested configuration
    client.set_config_override(
        "test-flag",
        json!({
            "theme": {
                "colors": {
                    "primary": "#ff0000"
                }
            }
        }),
        Some(ConfigOverrideOptions::new().with_merge(true)),
    ).unwrap();

    let result = client.evaluate("test-flag", None).await.unwrap();

    let config = result.configuration.unwrap();
    assert_eq!(config["theme"]["colors"]["primary"], "#ff0000");
    assert_eq!(config["theme"]["colors"]["secondary"], "#6c757d");
    assert_eq!(config["theme"]["fontSize"], 16);
}

#[tokio::test]
async fn test_get_all_overrides() {
    let config = Config::new("sdk_test_key");
    let client = FlagClient::new(config).unwrap();

    client.set_config_override("flag1", json!({ "config": "a" }), None).unwrap();
    client.set_config_override(
        "flag2",
        json!({ "config": "b" }),
        Some(ConfigOverrideOptions::new().with_merge(true)),
    ).unwrap();
    client.set_variation_override("flag3", "variant_a");

    let config_overrides = client.get_config_overrides();
    assert_eq!(config_overrides.len(), 2);
    assert!(config_overrides.contains_key("flag1"));
    assert!(config_overrides.contains_key("flag2"));

    let variation_overrides = client.get_variation_overrides();
    assert_eq!(variation_overrides.len(), 1);
    assert!(variation_overrides.contains_key("flag3"));
}
